Starting point for issue #132: update_item@action='delete' with value

eukaryote opened the pull request:
Changes to update_item on both the model and connection, as well as a unit test, to restore the correct behavior of allowing deletion of items from a set while still preserving the ability to delete without having to specify a value when the entire set should be removed.

This fixes a bug since 1.4.4 where if you had a user with a set-valued attribute like:

```
user = UserModel(1, nicknames={'nick1',  'nick2', 'DELETEME'})
user.save()
```

Then if you tried to remove a nickname from the set, with:

```
user.update_item('nicknames', value=['DELETEME'], action='delete')
```

The query sent to DynamoDB is as if no value were provided, and DynamoDB removes the nicknames attribute rather than just removing the item from the set.

eukaryote commented:
@jlafon This is an old pull request for #133 that has been updated and improved. Could you please review the bug report and pull request and let me know if there is anything else I can do to help move this along?

garrettheel commented:
Reopening with new default branch

behos commented on pynamodb/connection/base.py:
```diff
@@ -783,14 +783,14 @@ def update_item(self,
             attr_type, value = self.parse_attribute(value, return_type=True)
             action = update.get(ACTION)
-            if attr_type is None and action is not None and action.upper() != DELETE:
+            if attr_type is None and action is not None and (action.upper() != DELETE or (value is not None and not(isinstance(value, str) and value.upper() == 'NULL'))):
```
Can you extract this condition to a function? It's being repeated in multiple places.

eukaryote commented:
I've updated the changes related to update_item to work with the new expression syntax, and added a test for the new update method and made a similar fix for that case too.

Please review my changes carefully, as I'm not familiar enough with the code base to be sure my changes haven't broken something else that wasn't currently being tested in the test suite.

behos commented:
During my testing around this it seemed to me like the check done in my similar PR wasn't causing any regressions, and yours is much more conservative than that one :)

https://github.com/pynamodb/PynamoDB/pull/345/files

We've been running our own patched version of this code with no issues so far.

jl-DaDar commented:
i am not sure why but this does not work for me. I am on 3.0.1

jpinner-lyft commented:
Closing in favor of #374
